projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,344 @@
+            .map(lambda pair: pair[1])
+```
+Misc functions to plot the CDF of the submission delay.
+
+
+```python
+MAX_DELAY_S = 60 * 60 * 96.0
+HOUR_IN_S = 60 * 60.0
+
+def setup_plot(title, max_x, area_border_x=0.1, area_border_y=0.1):
+    plt.title(title)
+    plt.xlabel("Delay (hours)")
+    plt.ylabel("% of pings")
+
+    plt.xticks(range(0, int(max_x) + 1, 2))
+    plt.yticks(map(lambda y: y / 20.0, range(0, 21, 1)))
+
+    plt.ylim(0.0 - area_border_y, 1.0 + area_border_y)
+    plt.xlim(0.0 - area_border_x, max_x + area_border_x)
+
+    plt.grid(True)
+
+def plot_cdf(data, **kwargs):
+    sortd = np.sort(data)
+    ys = np.arange(len(sortd))/float(len(sortd))
+
+    plt.plot(sortd, ys, **kwargs)
+    
+def calculate_submission_delay(p):
+    created = datetime.fromtimestamp(p["meta/creationTimestamp"] / 1000.0 / 1000.0 / 1000.0)
+    received = datetime.fromtimestamp(p["meta/Timestamp"] / 1000.0 / 1000.0 / 1000.0)
+    sent = datetime.fromtimestamp(mktime_tz(parsedate_tz(p["meta/Date"]))) if p["meta/Date"] is not None else received
+    clock_skew = received - sent
+
+    return (received - created - clock_skew).total_seconds()
+```
+### Validate the ping payload
+Check that the payload section contains the right entries with consistent values.
+
+
+```python
+subset = get_pings_properties(update_pings, ["id",
+                                             "clientId",
+                                             "meta/creationTimestamp",
+                                             "meta/Date",
+                                             "meta/Timestamp",
+                                             "application/buildId",
+                                             "application/channel",
+                                             "application/version",
+                                             "environment/system/os/name",
+                                             "payload/reason",
+                                             "payload/targetBuildId",
+                                             "payload/targetChannel",
+                                             "payload/targetVersion",
+                                             "payload/previousBuildId",
+                                             "payload/previousChannel",
+                                             "payload/previousVersion"])
+```
+
+```python
+ping_success = subset.filter(lambda p: p.get("payload/reason") == "success")
+ping_ready = subset.filter(lambda p: p.get("payload/reason") == "ready")
+
+ping_success_count = ping_success.count()
+ping_ready_count = ping_ready.count()
+ping_count = ping_ready_count + ping_success_count
+```
+Quantify the percentage of duplicate pings we're receiving. We don't expect this value to be greater than ~1%, which is the amount we usually get from `main` and `crash`: as a rule of thumb, we threat anything less than 1% as *probably* well behaving.
+
```

Comment:
Maybe we should ensure `ping_count == subset.count()` (or is a least close. We should only have the two reasons)

Or do the payload validation against the whole deduped subset instead of the success-only subset.

Reply:
The payload validation/analysis for the ping type with `reason = ready `was performed [here](http://reports.telemetry.mozilla.org/post/projects/update_ping_ready_nightly_validation.kp) so I was trusting the data and restricting the validation to `reason = success` and just using the "ready" data for the last part of the validation below. Should I reference the other analysis or do you want me to add one of the checks you suggested?

Reply:
What I worry is seeing that the pings with doctype `update` are either `reason=ready` or `reason=success`. I think it's a valid question if we have `reason=justbecause` or something else that gets filtered out due to these filters.

Reply:
I see, let me add an assert for that.